rfd/0175-static-host-users.md

```diff
@@ -0,0 +1,216 @@
+---
+author: Andrew Burke (andrew.burke@goteleport.com)
+state: draft
+---
+
+# RFD 175 - Static Host Users
+
+## Required Approvers
+
+- Engineering: @rosstimothy && @lxea && @@espadolini
+
+## What
+
+Teleport nodes will be able to create host users statically, i.e. independently
+of a Teleport user creating one when SSHing with the current host user creation.
+
+## Why
+
+Host users can be created and used (potentially by third-party services) without
+a Teleport user needing to log in first.
+
+## Details
+
+### UX
+
+To create a static host user, an admin will create a `static_host_user` resource:
+
+```yaml
+# foo-dev.yaml
+kind: static_host_user
+metadata:
+    name: foo-dev
+spec:
+    login: foo
+    node_labels:
+        env: dev
+```
+
+Then create it with `tctl`:
+
+```code
+$ tctl create foo-dev.yaml
+```
+
+The user `foo` will eventually appear on nodes with label `env: dev` once the
+`foo-dev` resource makes it through the cache.
+
+To update an existing static host user, an admin will update update `foo-dev.yaml`,
+then update the resource in Teleport with `tctl`:
+
+```code
+$ tctl create -f foo-dev.yaml
+```
+
+### Resource
+
+We will add a new resource to Teleport called `static_host_user`. This resource defines
+a single host user, including groups, sudoers entitlements, uid, and gid, as well as labels
+to select specific nodes the user should be created on.
+
+```yaml
+kind: static_host_user
+metadata:
+    name: hostuser
+spec:
+    login: user1
+    # groups and sudoers are identical to their role counterparts
+    groups: [abc, def]
+    sudoers: [
+        # ...
+    ]
+    # same as from user traits
+    uid: "1234"
+    gid: "5678"
+    # same as allow rules in roles
+    node_labels:
+        # ...
+    node_labels_expression: # ...
+```
+
+```proto
+message StaticHostUser {
+    string kind = 1;
+    string sub_kind = 2;
+    string version = 3;
+    teleport.header.v1.Metadata metadata = 4;
+
+    StaticHostUserSpec spec = 5;
+}
+
+message StaticHostUserSpec {
+    string login = 1;
+    repeated string groups = 2;
+    repeated string sudoers = 3;
+    string uid = 4;
+    string gid = 5;
```

Comment:
Why are these strings?

Reply:
They're strings here because they're strings in [user traits](https://goteleport.com/docs/enroll-resources/server-access/guides/host-user-creation/#step-23-optional-configure-the-uid-and-gid-for-the-created-users), and I wanted to keep them the same.